<a href="https://colab.research.google.com/github/Daniil178/Opportunity_Cup2022/blob/main/ETL-proccess.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Реализация программы по обработке данных баанковских операций

In [20]:
# импортируем необходимые библиотеки
import pandas as pd
import json

In [21]:
# открытие файла и преобразование его в сплошную строку, т.к. функция json.loads 
# не может считать иной формат 
with open('transactions.json', "r") as f:
  s = f.read()
s = s.replace('\n', '')
s = s.replace('\t', '')

#загрузка строки, как json файла и снятие вложенности 'transactions'
trans = json.loads(s)
trans = trans['transactions']

#Запись в файл без обертки 'transactions', \n и \t
with open('trans.json', "w+") as f:
  f.write(json.dumps(trans))

In [22]:
# загрузка файла и изменение индексов (pandas автоматически преобразовывает в даты)
df = pd.read_json('trans.json')
df.columns = list(trans.keys())

In [ ]:
# Транспонирование матрицы для более удообного представление
df_T = df.T
lst_clients = list(set(list(df_T['client'])))

In [70]:
# Проверка таблицы
df_T

date                  card               account  \
459270924  2020-05-01T00:00:29  59649132026167121328  40817810000001139973   
459270943  2020-05-01T00:02:46  54236863445669263724  40817810000006468175   
459270968  2020-05-01T00:05:14  45930894029725313464  40817810000009142118   
459271057  2020-05-01T00:07:05  47550893546403715714  40817810000005559732   
459271064  2020-05-01T00:08:27  53036257457049008830  40817810000007495265   
...                        ...                   ...                   ...   
459275623  2020-05-05T21:33:12  48201610882535918830  40817810000009995390   
459275624  2020-05-05T21:34:46  58685860072058751139  40817810000009995391   
459275625  2020-05-05T21:36:07  52002457754706265318  40817810000009995392   
459275626  2020-05-05T21:37:48  46860025272700909519  40817810000009995393   
459275627  2020-05-05T21:39:29  56037470176508885939  40817810000009995394   

              account_valid_to   client        last_name first_name  \
459270924  2036-01-16T00:00:00  3-95179            Мисик     Сергей   
459270943  2033-11-15T00:00:00  8-33140          Ачентов      Антон   
459270968  2029-03-18T00:00:00  4-72043          Гуляков    Евгений   
459271057  2028-04-01T00:00:00  1-85045          Гуляков     Кирилл   
459271064  2037-10-26T00:00:00  5-77607            Мисик     Сергей   
...                        ...      ...              ...        ...   
459275623           2023-10-02  1-56145  Красносивенький      Ислам   
459275624           2033-06-21  1-29405         Павлихин       Гоар   
459275625           2017-05-28  8-44184         Абрамчик     Иоаким   
459275626           2017-07-21  8-44184         Абрамчик     Иоаким   
459275627           2014-01-21  8-44184         Абрамчик     Иоаким   

               patronymic        date_of_birth    passport  \
459270924      Николаевич  1938-06-25T00:00:00  7076445954   
459270943      Дмитриевич  1960-08-04T00:00:00  6403074379   
459270968  Константинович  1989-08-15T00:00:00  3354134791   
459271057        Петрович  1996-09-18T00:00:00  4783507057   
459271064        Олегович  1992-05-08T00:00:00  3195562178   
...                   ...                  ...         ...   
459275623         Ахматов           1952-03-13  4775599842   
459275624        Евсюшкин           1948-11-13  7932681937   
459275625        Свободин           1970-07-01  6144597998   
459275626        Свободин           1970-07-01  6144597998   
459275627        Свободин           1970-07-01  6144597998   

             passport_valid_to         phone   oper_type    amount  \
459270924  2022-11-09T00:00:00  +79497481039  Пополнение   31576.6   
459270943  2030-09-13T00:00:00  +79458324907  Пополнение   60041.1   
459270968  2021-06-24T00:00:00  +79953939977      Снятие  18474.46   
459271057  2023-01-08T00:00:00  +79698158422  Пополнение    915.26   
459271064  2032-09-27T00:00:00  +79181011177      Оплата  75843.91   
...                        ...           ...         ...       ...   
459275623           2030-11-11  +73497185681  Пополнение  33456.18   
459275624           2034-03-02  +76101544386      Оплата  60893.22   
459275625           2022-01-12  +77469130963      Оплата  17421.05   
459275626           2027-04-14  +77469130963  Пополнение  27723.81   
459275627           2031-11-10  +77469130963      Снятие  84481.15   

          oper_result  terminal terminal_type                    city  \
459270924       Отказ  POS43792           POS      Славянск-на-Кубани   
459270943       Отказ  POS75616           POS                Пустошка   
459270968     Успешно  POS28311           POS  Петровск-Забайкальский   
459271057     Успешно  ATM39555           ATM          Краснослободск   
459271064       Отказ  ATM16730           ATM                Черкесск   
...               ...       ...           ...                     ...   
459275623       Отказ  POS43792           POS                    Зима   
459275624       Отказ  POS43792           POS                   Вы

***Пример транзакции:***

{'date': '2020-05-01T00:00:29',

 'card': '59649132026167121328',
 
 'account': '40817810000001139973',
 
 'account_valid_to': '2036-01-16T00:00:00',
 
 'client': '3-95179',
 
 'last_name': 'Мисик',
 
 'first_name': 'Сергей',
 
 'patronymic': 'Николаевич',
 
 'date_of_birth': '1938-06-25T00:00:00',
 
 'passport': 7076445954,
 
 'passport_valid_to': '2022-11-09T00:00:00',
 
 'phone': '+79497481039',
 
 
 'oper_type': 'Пополнение',
 
 'amount': 31576.6,
 
 'oper_result': 'Отказ',
 
 'terminal': 'POS43792',
 
 'terminal_type': 'POS',
 
 'city': 'Славянск-на-Кубани',
 
 'address': 'Славянск-на-Кубани, ул. Клецкая, д. 86'}

In [56]:
def diff_city(df): #паттерн 1: посещение городов за малый промежуток времени
  lst_trans = list(df.index)
  lst_minutes = []
  lst_date = list(df['date'])
  lst_city = list(df['city'])
  for i in range(len(lst_date)):
    lst_minutes.append(int(lst_date[i].split('T')[1].split(':')[1]) + 60 * int(lst_date[i].split('T')[1][:2]) + 24 * 60 * (int(lst_date[i].split('T')[0][-1]) - 1)) #выделение минут
  for i in range(1, len(lst_date)):
    if (lst_minutes[i] - lst_minutes[i-1]) < 30 and lst_city[i] != lst_city[i-1]:  # проверка на условие (за 30 минут приехал в другой город)
      return lst_trans[i] # номер транзакции
    else:
      return None

def not_valid_account(df): #паттерн 2: не действующий аккаунт
  lst_trans = list(df.index)
  lst_month = []
  lst_date = list(df['date'])
  lst_valid = list(df['account_valid_to'])
  for i in range(len(lst_date)):
    lst_month.append(int(lst_date[i].split('-')[0]) * 12 + int(lst_date[i].split('-')[1])) # список с месяцами операций
    lst_valid[i] = int(int(lst_valid[i].split('-')[0]) * 12 + int(lst_valid[i].split('-')[1])) # срок валидности аккаунта
    if lst_valid[i] - lst_month[i] < -6: # проверка на просрочку в пол года (первые пол года человек может воспользоваться аккаунтом)
      return lst_trans[i] # номер транзакции
    else:
      return None

def not_valid_passport(df): #паттерн 3: недействительный паспорт
  lst_trans = list(df.index)
  lst_month = []
  lst_date = list(df['date'])
  lst_valid = list(df['passport_valid_to'])
  for i in range(len(lst_date)):
    lst_month.append(int(lst_date[i].split('-')[0]) * 12 + int(lst_date[i].split('-')[1])) # список с месяцами операций
    lst_valid[i] = int(int(lst_valid[i].split('-')[0]) * 12 + int(lst_valid[i].split('-')[1])) # срок валидности паспорта
    if lst_valid[i] - lst_month[i] < -3: # проверка на просрочку (первые 3 месяца человек может пользоваться картой)
      return lst_trans[i] # номер транзакции
    else:
      return None

def fast_trans(df): # паттерн 4: совершение транзакций с разных карт за малый промежуток времени
  lst_trans = list(df.index)
  lst_minute = []
  lst_date = list(df['date'])
  lst_card = list(df['card'])
  for i in range(len(lst_date)):
    lst_minute.append(int(lst_date[i].split('T')[1].split(':')[1]) + 60 * int(lst_date[i].split('T')[1][:2]) + 24 * 60 * (int(lst_date[i].split('T')[0][-1]) - 1)) # время операции в минутах
  for i in range(1, len(lst_date)):
    if (lst_minute[i] - lst_minute[i-1]) < 2 and lst_card[i] != lst_card[i-1]: # проверка на условие (за 2-3 минуты человек может сменить карту для использования)
      return lst_trans[i] # номер транзакции
    else:
      return None

def period(df): #паттерн 5: проверка транзакций на периодичность
  lst_trans = list(df.index)
  lst_sec = []
  lst_diff = []
  lst_date = list(df['date'])
  for i in range(len(lst_date)): #ниже время операции в секундах
    lst_sec.append(int(lst_date[i].split('T')[1].split(':')[-1]) + 60 *int(lst_date[i].split('T')[1].split(':')[1]) + 60 * 60 * int(lst_date[i].split('T')[1][:2]) + 24 * 60 * 60 * (int(lst_date[i].split('T')[0][-1]) - 1))
  for i in range(1, len(lst_trans)):
    lst_diff.append(lst_sec[i] - lst_sec[i-1]) #вычисление разности времени транзакций
  for i in range(1, len(lst_diff)):
    if lst_diff[i] - lst_diff[i-1] < 0.2 * lst_diff[i-1]: # проверка условия (с учетом погрешности)
      return lst_trans[i] # номер транзакции
    else:
      return None

def average(df): # паттерн 6: проверка на сумму операции
  lst_trans = list(df.index)
  lst_money = list(df['amount'])
  if len(lst_money) > 2: # для 2х и меньше операций нет смысла вычислять эти данные (слишком маленькая выборка) 
    aver = sum(lst_money)/len(lst_money) # среднее значение суммы операции по клиенту
    deviation = 0
    for x_i in lst_money:
      deviation += (x_i - aver)**2
    deviation = (deviation / len(lst_money)) ** 0.5 # вычисляем несммещенную оценку дисперсии
    for i in range(len(lst_money)):
      if abs(lst_money[i] - aver) > deviation: # проверка условия
        return lst_trans[i] # номер транзакции
      else:
         return None
  else:
    return None

def time_to_sleep(df): # паттерн 7: проверка на совершение операции в необычное время с учетом возраста
  lst_trans = list(df.index)
  lst_age = list(df['date_of_birth'])
  lst_time = list(df['date'])
  for i in range(len(lst_age)):
    lst_age[i] = int(lst_time[i].split('-')[0]) - int(lst_age[i].split('-')[0]) # вычисление возраста клиента
    lst_time[i] = int(lst_time[i].split('T')[1][:2]) # вычисление времени операции в часах
    if (30 < lst_age[i] < 50) and not (7 < lst_time[i] < 23): # проверки условий
      return lst_trans[i]
    elif (50 < lst_age[i] < 60) and not (7 < lst_time[i] < 22):
      return lst_trans[i]
    elif (lst_age[i] >= 60) and not (8 < lst_time[i] < 22):
      return lst_trans[i] # номер транзакции
    else:
      return None

def result_frods(patts): # паттерн 8: итоговые фроды, с учетом паттернов и их весов
  score = [2, 1, 1, 2, 2, 1, 1] # список весов каждого паттерна
  res_patts = []
  score_patt = {}
  all_frods = list(set(sum_list(patts))) # все операции, рассматриваемые, как фроды
  for trans in all_frods:
    score_patt[trans] = 0 # инициализация словаря
  for patt in all_frods:
    for i in range(len(patts)):
      if patt in patts[i]:
        score_patt[patt] += score[i] # добавление веса к транзакции в зависимости от принадлежности к паттерну
    if score_patt[patt] > 3:
      res_patts += [patt] # добавление в итоговый список
  return res_patts

In [12]:
def sum_list(A): # конкатенация списков, содержащихся в облее крупном
  B = []
  for i in range(len(A)):
    B += A[i]
  return B

In [60]:
# блок обработки информации
num_patts = 7 # общее число базовых паттернов
patts = [[] for i in range(num_patts)]
func = [diff_city, not_valid_account, not_valid_passport, fast_trans, period, average, time_to_sleep] # функции фильтрации
for client in lst_clients:
  df = df_T[df_T['client'] == client]
  for i in range(num_patts):
    p = func[i](df)
    if p != None:
      patts[i].append(p)

# добавление итогового паттерна
num_patts += 1
patts.append(result_frods(patts))

# вывод промежуточных результатов программы
for i in range(num_patts):
  print(f'patt {i+1} - number of transactions: {len(patts[i])}')
print(f'all frods: {len(sum_list(patts))}, unique: {len(set(sum_list(patts)))}')

patt 1 - number of transactions: 184
patt 2 - number of transactions: 260
patt 3 - number of transactions: 241
patt 4 - number of transactions: 422
patt 5 - number of transactions: 301
patt 6 - number of transactions: 200
patt 7 - number of transactions: 405
all frods: 2013, unique: 1419


In [62]:
#подготовка данных к загрузке в excel
tab = []
for i in range(num_patts):
  tab.append([i+1, patts[i]])

In [63]:
# создание DataFrame на основе наших данных
table = pd.DataFrame(tab, columns = ['№ Паттерна', 'Номера транзакций, попадающих под паттерн'])

In [66]:
# проверка таблицы
table

№ Паттерна          Номера транзакций, попадающих под паттерн
0           1  [459275276, 459273992, 459275259, 459273720, 4...
1           2  [459273855, 459275055, 459274384, 459274701, 4...
2           3  [459275330, 459274327, 459274254, 459275393, 4...
3           4  [459275276, 459273867, 459275331, 459275259, 4...
4           5  [459275276, 459275331, 459275259, 459271317, 4...
5           6  [459275330, 459271178, 459275021, 459273760, 4...
6           7  [459273866, 459273991, 459273855, 459275055, 4...
7           8  [459273820, 459274883, 459274743, 459273901, 4...

In [65]:
# загрузка в excel
table.to_excel('ответ.xlsx', sheet_name='Транзакции', index=False)